### ISSUE TITLE
TypedDict does not take in account Required and NotRequired

### ISSUE THREAD

**Issue by dgellow** [microsoft/pyright#3301]
`Required` and `NotRequired` from [PEP-655](https://peps.python.org/pep-0655/) do not seem to be taken in account when type checking `TypeDict`.

## Expected behavior

I expect pyright/pylance to not report a key as missing if it is marked as `NotRequired` with a `TypedDict, total=False`. And a missing key marked as `Required` with a `TypedDict, total=True` to be reported as a type error.

## Actual behavior

Both `Required` and `NotRequired` seem to be ignored.

File `_typing.py`:
```python
import sys

# PEP 655 – Marking individual TypedDict items as required or potentially-missing
# https://peps.python.org/pep-0655/
#
# Unpack: PEP 646 – Variadic Generics
# https://peps.python.org/pep-0646/
if sys.version_info >= (3, 11):
    from typing import Required, NotRequired, Unpack
else:
    from typing_extensions import Required, NotRequired, Unpack

# PEP 589 – TypedDict: Type Hints for Dictionaries with a Fixed Set of Keys
# https://peps.python.org/pep-0589/
#
# PEP 586 – Literal Types
# https://peps.python.org/pep-0586/
if sys.version_info >= (3, 8):
    from typing import TypedDict, Literal
else:
    from typing_extensions import TypedDict, Literal

```

File `example.py`:
```python
from ._typing import Unpack, TypedDict, Literal, Required, NotRequired

class MyKwargs(TypedDict):
    foo: NotRequired[Literal["hello", "nopnop"]]
    bar: NotRequired[int]

MyKwargs() # 👈 Here a type error 'Arguments missing for parameters "foo", "bar"' is reported

```

## Screenshot of the error

![image](https://user-images.githubusercontent.com/2451004/161973389-38cc6e0c-bbf2-4647-bb65-349a6cec7a2b.png)

**VS Code extension or command-line**

Pylance version v2022.3.4, from vscode.

**Comment by dgellow** [comment]
I found this previous discussion: https://github.com/microsoft/pylance-release/issues/2469#issuecomment-1066829917.

> This is expected behavior. The problem is that you're re-exporting Required and NotRequired from a local file (utils.py). These are special forms that pyright (the type checker that underlies pylance) needs to handle early in its analysis, so it can't deal with this form of aliasing. You must import them directly from typing or typing_extensions if you want them to work with pyright.

That still feel like a bug to me, but this issue should maybe be reclassified as a feature request.

**Comment by erictraut** [comment]
This is as designed. Please import these special forms directly from `typing_extensions` or `typing`.

**Comment by dgellow** [comment]
@erictraut Just to clarify one thing, should we expect this for everything exported by `typing_extensions`, or is it specific to `Required` and `NotRequired`? I'm not sure I understand what you mean by "special forms" here, or how I can identify what is or not considered a special form.

For example if I use Python 3.7, would you also recommend `TypedDict` to be imported directly from `typing_extensions` and not re-exported by a utility file?

I checked pyright documentation but did not find details on this, but I may have missed it.

**Comment by erictraut** [comment]
If you look in `typing.pyi`, you'll see the term `SpecialForm`. This terminology is used in PEP 484 as well. It refers to symbols that have special meaning to type checkers.

This limitation applies specifically to special forms that need to be understood by pyright's parser and binder, the two earliest stages that occur before type evaluation. 

It applies to the following special forms:
* Annotated
* Final
* TypeAlias
* ClassVar
* Required
* NotRequired
* Literal

What most of these have in common is that they are forms that do not specify a type but rather specify some ancillary (non-type) property of an attribute. In most languages, keywords or dedicated syntax would have been added to accommodate these concepts. In Python, there is an aversion to adding type-related keywords and syntax, so these concepts were unfortunately shoehorned into the type annotation syntax. This makes them difficult for type checkers to deal with because they're not really type annotations. In pyright's case, we need to understand the meaning of these symbols within the binder, but the binder cannot perform arbitrary type evaluation. Its output is required for the type evaluator. 

I haven't included this in any formal documentation because, frankly, I didn't think anyone would ever attempt to import and re-export these symbols from within their own code. Please import them directly from `typing_extension`, and it won't be a problem.

**Comment by dgellow** [comment]
Thank you very much for such a good amount of details! That's very helpful 😃

**Comment by hmc-cs-mdrissi** [comment]
The only reason I've occasionally had to do weird things for depending on typing-extensions is some packages have upper bound version limits. typing-extensions<4 is one I've come across in a couple packages as they guessed major version change would be an issue and it's blocked my ability to use Required/NotRequired from typing extensions normally. Usually I still import typing-extensions and do `if TYPE_CHECKING` workarounds. This isn't really a typing problem, but python packaging issue of dealing with conflicts is messy and is my guess why some libraries try to avoid dependencies.

Thankfully it is rare. I think there's 3 packages I've encountered so far that had upper bound limits on typing extensions and I had to either upstream a pr/wait for a new release. I'm currently debating forking one dependency solely to change it's typing extensions bound.

**Comment by dgellow** [comment]
> I don't see why people jump through hoops with version checks so much.

Just my 2 cents.

I'm myself not that much experienced with Python and found out about `typing_extensions` by reading answers on stackoverflow and other forums/doc pages mentioning the following pattern:
```python
if sys.version_info >= (3, 11):
    from typing import SomehingToImport
else:
    from typing_extensions import SomethingToImport
```

Here `typing_extensions` is presented as a kind of "support library for older versions of pythons" and it seems to be reasonable to do a version check before using it. So I assumed that was the general accepted/correct way to handle the situation and repeated it, and I would guess a lot of people do the same without questioning it too much. But in fact, if I now understand correctly, it is an actual drop-in replacement of `typing` **and** has backported type hints for older versions of Python and experimental stuff. 

I just read a bit of the implementation, following this thread, and see that `typing_extensions` already does version checks internally (I mean, that's the whole point of the library, right? 🙈 ), which makes the whole thing redundant and `from typing_extensions import Whatever` would perfectly replace all my imports of the `typing` package in my project. But that wasn't really obvious until I had a reason to look into it.

**Comment by JelleZijlstra** [comment]
> it is an actual drop-in replacement of typing and has backported type hints for older versions of Python and experimental stuff

Not quite, it doesn't have things that have been in `typing` since the beginning (e.g. there is no `typing_extensions.Any`). But anything that was ever in typing_extensions we'll keep.
